### PR TITLE
Multi amplitude embedding

### PR DESCRIPTION
**Pull request by KetpuntoG**
**Context:** PennyLane does not allow multiple state preparations at the beginning of a circuit. [(#1140)](https://github.com/PennyLaneAI/pennylane/issues/1140)

**Description of the Change:** The possibility of incorporating several AmplitudeEmbedding in the same circuit has been added as long as they do not coincide in any of the application qubits. Example:

```python
import numpy as np
import pennylane as qml

dev = qml.device("default.qubit", wires=5)
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires = 0)
    qml.templates.AmplitudeEmbedding([1,1,1,1], wires = [1,2], normalize = True)
    qml.templates.AmplitudeEmbedding([1,1,-1,-1], wires = [3,4], normalize = True)
    qml.CSWAP(wires = [0,1,3])
    qml.CSWAP(wires = [0,2,4])
    qml.Hadamard(wires = 0)
    return qml.probs(wires = 0)

circuit()[0]
```
>>>  tensor(0.5, requires_grad=True)

**Benefits:** It facilitates the application to examples such as the swap test to calculate the inner product of two given states.

**Possible Drawbacks:** Currently it only works with AmplitudeEmbedding and Default.qubit. It is not possible to apply to BasisState.

**Related GitHub Issues:** [(#1140)](https://github.com/PennyLaneAI/pennylane/issues/1140)

### CODE REVIEW

File: tests/devices/test_default_qubit_tf.py [hunk 1]
```diff
@@ -313,7 +313,7 @@ def test_invalid_state_prep(self):
         """Test that an exception is raised if a state preparation is not the
         first operation in the circuit."""
         dev = DefaultQubitTF(wires=2)
-        state = np.array([0, 12])
+        state = np.array([0, 1])
```
Comment (KetpuntoG) [inline tests/devices/test_default_qubit_tf.py:316]: The error: 

```python
        with pytest.raises(
            qml.DeviceError,
            match=r"cannot be used after other Operations have already been applied",
        ):
            dev.apply([qml.PauliZ(0), qml.QubitStateVector(state, wires=[0])])
```
is detected but the error that the state is not normalized is displayed first. That is why I have changed the test. I don't know if this is possible

**Comment by KetpuntoG** [comment]
[sc-11240]

File: tests/templates/test_embeddings/test_amplitude.py [hunk 2]
```diff
@@ -359,3 +359,49 @@ def test_torch(self, tol):
         res2 = circuit2(features)
 
         assert qml.math.allclose(res, res2, atol=tol, rtol=0)
+
+
+def test_multiple_amplitudes():
```
Comment (antalszava) [inline tests/templates/test_embeddings/test_amplitude.py:364]: It's fair game to create a separate test class here (just like `TestInterfaces`). :slightly_smiling_face: That way further test cases can also be added as separate methods.

File: tests/templates/test_embeddings/test_amplitude.py [hunk 3]
```diff
@@ -359,3 +359,49 @@ def test_torch(self, tol):
...
+    """Tests the use of many AmplitudeEmbedding."""
+
+    tol = 10e-10
+    dev = qml.device("default.qubit", wires=7)
+
+    @qml.qnode(dev)
+    def circuit():
+        qml.Hadamard(wires=0)
+        qml.templates.AmplitudeEmbedding([1, 1, 1, 1, 0, 0, 0, 0], wires=[1, 2, 3], normalize=True)
```
Comment (antalszava) [inline tests/templates/test_embeddings/test_amplitude.py:373]: Maybe it's worth checking if we have an operation in between two AmplitudeEmbedding operations.

**Review by antalszava — COMMENTED** [review]
Hi @KetpuntoG, a great one with working out the logic here! :muscle: :slightly_smiling_face: Left a couple of comments, it seems that `AmplitudeEmbedding` might yield different cases to using `QubitStateVector` at times. It would be useful to have the test cases for `AmplitudeEmbedding` also for `QubitStateVector` e.g., in the `tests/devices/test_default_qubit.py` file.

File: pennylane/devices/default_qubit.py [hunk 4]
```diff
@@ -202,21 +202,49 @@ def apply(self, operations, rotations=None, **kwargs):
         rotations = rotations or []
 
         # apply the circuit operations
+        wires_visited = set()
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/devices/default_qubit.py:205]: This is something that struck me as odd when I saw it but currently pennylane works like this. What the circuit does is to sort it and put the embedding operations first. If you paint it you can see that the Hadamard gate is after the AmplitudeEmbedding. I think it would be more accurate to throw an error instead of ordering it but I don't know if it has another reason that I don't know.

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/devices/default_qubit.py:205]: This was an intentional restriction at the time; the reason being was that PL originally did not support mixed states, and a mixed state would result if you applied an embedding to only some of the wires in the middle of the circuit!

However something we can definitely reconsider now that we have mixed state simulator support

File: pennylane/devices/default_qubit.py [hunk 5]
```diff
@@ -202,21 +202,49 @@ def apply(self, operations, rotations=None, **kwargs):
...
+        input_vectors, input_wires = [], []
+        has_qubit_vector_state = False
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/devices/default_qubit.py:207]: Oops, that's right! this boolean is not necessary 😄

File: pennylane/devices/default_qubit.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/devices/default_qubit.py:205]: What I find curious is that AmpllitudeEmbedding and QubitStateVector work differently. What is the difference we want to show between them @josh146 ?

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/devices/default_qubit.py:205]: How do you mean by work differently, @KetpuntoG? 🙂

[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/devices/default_qubit.py:205]:
```python
import numpy as np
import pennylane as qml

dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires = 0)
    qml.AmplitudeEmbedding([1,0,0,0], wires = [0,1])
    #qml.QubitStateVector([1,0,0,0], wires = [0,1])
    return qml.expval(qml.PauliZ(0))

circuit()
```

In this case only AmplitudeEmbedding will work.

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/devices/default_qubit.py:205]: Ah, this is not intentional 😆 The `AmplitudeEmbedding` template is getting around our check that state preparation only occurs at the beginning of the circuit!

[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/devices/default_qubit.py:205]: Ah ok hahah makes sense.  Anyway I would change it to one that checks that no qubits have been used, which is kind of the idea I'm doing with this new AmplitudeEmbedding.

**Comment by KetpuntoG** [comment]
In the end I have removed the restriction that QubitStateVector can only be defined at the beginning. This way, we can use several QubitStateVectors and AmplitudeEmbedding indistinctly. On the other hand, this solution for the moment is exclusive for AmplitudeEmbedding, it would be pending to be able to make multi BasisState and that they can join together in a single AmplitudeEmbedding.

File: pennylane/devices/default_qubit.py [hunk 6]
```diff
@@ -620,17 +647,25 @@ def _apply_state_vector(self, state, device_wires):
...
         Args:
-            state (array[complex]): normalized input state of length
-                ``2**len(wires)``
-            device_wires (Wires): wires that get initialized in the state
+            state (array[array[complex]]): normalized input states
+            device_wires (array[Wires]): wires that get initialized in the state
         """
-
+        states = state
+        if type(device_wires) != list:
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:654]: Two points here:
1. Would we allow lists here or arrays? The docstring mentions the latter. If we allow multiple options (e.g., even tuples), then we could potentially check for `Sequence` here;
2. It might be better to check with `isinstance`, see [here for some details](https://stackoverflow.com/a/1549854/12899253).

File: pennylane/devices/default_qubit.py [hunk 7]
```diff
@@ -620,17 +647,25 @@ def _apply_state_vector(self, state, device_wires):
...
+            states = [states]
+            device_wires = [device_wires]
+        state = states[0]
+        wires = device_wires[0]
+        state = self._asarray(state, dtype=self.C_DTYPE)
+        # the final state is calculated as the tensor product of each of the states.
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:660]:
```suggestion
        # The final state is calculated as the tensor product of each of the states.
```

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @KetpuntoG, this is getting there :rocket: :slightly_smiling_face: Left some comments, they are mostly questions and minor adjustments.

What I think would be important to have is more testing, for example for `QubitStateVector` too, as the main change in this PR affects this operation.

Further to this, it would be great to check that applying multiple `QubitStateVector`/`AmplitudeEmbedding` operations works well when none of the qubits is used more than once with further in-built devices (`default.mixed` and each framework variant of `default.qubit`). It's okay if there are known cases that behave differently (such as the catch you had). The idea would be to make sure that the functionality that this PR allows (multiple AmplitudeEmbedding & QubitStatevector) works great with built-in devices. There are also some edge cases for each of the framework variant devices in the `tests/devices` in the test folder (e.g., multiple cases of JIT support with JAX).

**Comment by mariaschuld** [comment]
Hey @KetpuntoG, I just stumbled across this, and if you don't mind I'll drop a few thoughts here.

Allowing for `QubitStatePreparation` (and hence `AmplitudeEmbedding`, which adds some preprocessing) to be callable on different subsystems has been something we discussed many times before, and always shied away from because it adds so much specific logic we will have to maintain going forward - so this is not a trivial PR :).
 
I'm just curious about one issue, but struggle to understand it immediately from the code (already a long day!). Say you use  `QubitStatePreparation` to set subsystems 0,1 to a normalised state and then 2, 3, 4 to another normalised state. How would is the final full state vector renormalised? Would your implementation weigh the two AmplitudeEmbeddings 50%-50% or according to the number of qubits? In other words, is the desired behaviour clear here?

And then, to play devil's advocate: Why can't the user apply one big `AmplitudeEmbedding` or `QubitStatePreparation`, and fine tune the behaviour? 

Sorry to come in with these comments, hope they help to think about the changes!

**Comment by KetpuntoG** [comment]
Everything you say makes perfect sense @mariaschuld ! Regarding the 50%-50%. Being two independent normalized states, all I do is calculate the tensor product of both states. Look at it as two different circuits, if we now put them together, the current state is the tensor product of both.

On the other hand, you could perfectly well put everything in a single AmplitudeEmbedding, my goal was to avoid the user having to do this tensor product in case he does not know the concept. But I understand that it is still a lot of change in the logic for this functionality that the user could learn quickly.

**Comment by mariaschuld** [comment]
Ah I see, the tensor product makes total sense I guess!

About the other concerns, I'm sure you'll come to the right decision - it is so hard to find the balance between providing nice features and keeping the code base maintainable. I think the current big operator class refactor (for which each piece of logic creates another nightmare) makes me very biased :)

**Comment by KetpuntoG** [comment]
Thank you Maria, I think all this that you tell me! 😊

**Comment by KetpuntoG** [comment]
It has been decided to make a very different approach through transformations: [(#1933)](https://github.com/PennyLaneAI/pennylane/pull/1933)

I wanted to do it in a new branch so as not to overwrite this PR which I think is interesting.
